File: typescript/src/ancilis/activation/loader.ts
/** Loads overlay profiles and certification profiles from shared JSON data files. */

import { readFileSync, readdirSync, existsSync } from "node:fs";
import { join } from "node:path";
import { sharedPathFrom } from "../shared-path.js";
import { type PluginContext, PluginRegistry } from "../plugins/index.js";

const SHARED_DIR = sharedPathFrom(import.meta.url);
const OVERLAYS_DIR = join(SHARED_DIR, "overlays");
const CERTIFICATIONS_DIR = join(OVERLAYS_DIR, "certifications");
const CONTROLS_DIR = join(SHARED_DIR, "controls");
const CLASSIFICATIONS_FILE = join(SHARED_DIR, "classifications", "taxonomy.json");

export interface LoadOverlayProfilesOptions {
  readonly pluginRegistry?: PluginRegistry;
  readonly pluginConfigs?: Readonly<Record<string, Readonly<Record<string, unknown>>>>;
  readonly warnings?: string[];
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isStringList(value: unknown): value is string[] {
  return Array.isArray(value) && value.every((item) => typeof item === "string");
}

function warnOverlay(pluginName: string, detail: string, warnings?: string[]): void {
  const message = `Skipping Ancilis plugin overlay ${pluginName}: ${detail}`;
  warnings?.push(message);
}

function validatedPluginOverlayProfile(
  pluginName: string,
  rawProfile: unknown,
  warnings?: string[],
): Record<string, unknown> | null {
  if (!isRecord(rawProfile)) {
    warnOverlay(pluginName, "overlay profile must be a mapping", warnings);
    return null;
  }

  const overlayId = rawProfile.id;
  if (
    typeof overlayId !== "string"
    || !overlayId.startsWith("plugin:")
    || overlayId.slice("plugin:".length).trim().length === 0
  ) {
    warnOverlay(pluginName, "overlay id must be explicit and namespaced as plugin:<name>", warnings);
    return null;
  }

  for (const fieldName of ["name", "trigger_type"] as const) {
    const value = rawProfile[fieldName];
    if (typeof value !== "string" || value.trim().length === 0) {
      warnOverlay(pluginName, `overlay profile missing required '${fieldName}' field`, warnings);
      return null;
    }
  }

  for (const fieldName of ["triggered_by", "applicable_data_types"] as const) {
    const value = rawProfile[fieldName];
    if (value !== undefined && value !== null && !isStringList(value)) {
      warnOverlay(pluginName, `overlay profile field '${fieldName}' must be a list of strings`, warnings);
      return null;
    }
  }

  for (const fieldName of ["control_adjustments", "evidence_requirements", "controls"] as const) {
    const value = rawProfile[fieldName];
    if (value !== undefined && value !== null && !isRecord(value)) {
      warnOverlay(pluginName, `overlay profile field '${fieldName}' must be a mapping`, warnings);
      return null;
    }
  }

  const controlAdjustments = rawProfile.control_adjustments;
  if (isRecord(controlAdjustments)) {
    for (const [controlId, adjustment] of Object.entries(controlAdjustments)) {
      if (!isRecord(adjustment)) {
        warnOverlay(pluginName, `overlay profile field 'control_adjustments.${controlId}' must be a mapping`, warnings);
        return null;
      }
    }
  }

  const evidenceRequirements = rawProfile.evidence_requirements;
  if (isRecord(evidenceRequirements)) {
    for (const [controlId, requirements] of Object.entries(evidenceRequirements)) {
      if (!isStringList(requirements)) {
        warnOverlay(pluginName, `overlay profile field 'evidence_requirements.${controlId}' must be a list of strings`, warnings);
        return null;
      }
    }
  }

  const controls = rawProfile.controls;
  if (isRecord(controls)) {
    for (const [controlId, controlData] of Object.entries(controls)) {
      if (!isRecord(controlData)) {
        warnOverlay(pluginName, `overlay profile field 'controls.${controlId}' must be a mapping`, warnings);
        return null;
      }
      const controlRequirements = controlData.evidence_requirements;
      if (controlRequirements !== undefined && controlRequirements !== null && !isStringList(controlRequirements)) {
        warnOverlay(pluginName, `overlay profile field 'controls.${controlId}.evidence_requirements' must be a list of strings`, warnings);
        return null;
      }
    }
  }

  const retentionDays = rawProfile.evidence_retention_minimum_days;
  if (retentionDays !== undefined && (!Number.isInteger(retentionDays) || typeof retentionDays !== "number")) {
    warnOverlay(pluginName, "overlay profile field 'evidence_retention_minimum_days' must be an integer", warnings);
    return null;
  }

  const humanOversightRequired = rawProfile.human_oversight_required;
  if (humanOversightRequired !== undefined && typeof humanOversightRequired !== "boolean") {
    warnOverlay(pluginName, "overlay profile field 'human_oversight_required' must be a boolean", warnings);
    return null;
  }

  return { ...rawProfile };
}

function mergePluginOverlayProfiles(
  profiles: Map<string, Record<string, unknown>>,
  options: LoadOverlayProfilesOptions,
): void {
  const pluginRegistry = options.pluginRegistry;
  if (!pluginRegistry) return;

  for (const record of pluginRegistry.compatible("overlay")) {
    const pluginName = record.name;
    const loadedPlugin = record.plugin;
    if (!isRecord(loadedPlugin)) {
      warnOverlay(pluginName, "has no plugin object and was skipped", options.warnings);
      continue;
    }
    if (typeof loadedPlugin.loadOverlayProfile !== "function") {
      warnOverlay(pluginName, "does not implement the overlay plugin contract", options.warnings);
      continue;
    }

    const context: PluginContext = {
      sdkVersion: pluginRegistry.sdkVersion,
      config: options.pluginConfigs?.[pluginName] ?? {},
    };

    let rawProfile: unknown;
    try {
      rawProfile = loadedPlugin.loadOverlayProfile(context);
    } catch (error: unknown) {
      warnOverlay(
        pluginName,
        `failed to load overlay profile: ${(error as Error).message ?? String(error)}`,
        options.warnings,
      );
      continue;
    }

    const profile = validatedPluginOverlayProfile(pluginName, rawProfile, options.warnings);
    if (!profile) continue;

    const overlayId = profile.id as string;
    if (profiles.has(overlayId)) {
      warnOverlay(
        pluginName,
        `overlay id '${overlayId}' collides with an existing overlay and was skipped`,
        options.warnings,
      );
      continue;
    }
    profiles.set(overlayId, profile);
  }
}

export function loadOverlayProfiles(options: LoadOverlayProfilesOptions = {}): Map<string, Record<string, unknown>> {
  const profiles = new Map<string, Record<string, unknown>>();
  const files = readdirSync(OVERLAYS_DIR).filter(f => f.endsWith(".json")).sort();
  for (const file of files) {
    const data = JSON.parse(readFileSync(join(OVERLAYS_DIR, file), "utf-8"));
    profiles.set(data.id, data);
  }
  mergePluginOverlayProfiles(profiles, options);
  return profiles;
}

export function loadCertificationProfile(certId: string): Record<string, unknown> | null {
  const path = join(CERTIFICATIONS_DIR, `${certId}.json`);
  if (!existsSync(path)) {
    return null;
  }
  const data = JSON.parse(readFileSync(path, "utf-8"));
  if (!("version" in data)) {
    return null;
  }
  return data;
}

export function loadCertificationProfiles(certIds: string[]): Map<string, Record<string, unknown>> {
  const profiles = new Map<string, Record<string, unknown>>();
  for (const cid of certIds) {
    const profile = loadCertificationProfile(cid);
    if (profile !== null) {
      profiles.set(cid, profile);
    }
  }
  return profiles;
}

export function loadControlDefinitions(): Map<string, Record<string, unknown>> {
  const controls = new Map<string, Record<string, unknown>>();
  const files = readdirSync(CONTROLS_DIR).filter(f => f.endsWith(".json")).sort();
  for (const file of files) {
    const data = JSON.parse(readFileSync(join(CONTROLS_DIR, file), "utf-8"));
    controls.set(data.id, data);
  }
  return controls;
}

export function loadTaxonomy(): Record<string, unknown> {
  return JSON.parse(readFileSync(CLASSIFICATIONS_FILE, "utf-8"));
}

File: python/src/ancilis/activation/loader.py
"""Loads overlay profiles and certification profiles from shared JSON data files."""

from __future__ import annotations

import json
import logging
from collections.abc import Mapping
from typing import Any, cast

from ancilis._shared import shared_path
from ancilis.plugins import PluginContext, PluginRegistry

logger = logging.getLogger("ancilis.activation")

SHARED_DIR = shared_path()
OVERLAYS_DIR = SHARED_DIR / "overlays"
CERTIFICATIONS_DIR = OVERLAYS_DIR / "certifications"
CONTROLS_DIR = SHARED_DIR / "controls"
CLASSIFICATIONS_FILE = SHARED_DIR / "classifications" / "taxonomy.json"


def load_overlay_profiles(
    *,
    plugin_registry: PluginRegistry | None = None,
    plugin_configs: Mapping[str, Mapping[str, Any]] | None = None,
    warnings: list[str] | None = None,
) -> dict[str, dict[str, Any]]:
    """Load built-in and optional plugin overlay profiles."""
    profiles: dict[str, dict[str, Any]] = {}
    for path in sorted(OVERLAYS_DIR.glob("*.json")):
        data = json.loads(path.read_text())
        profiles[data["id"]] = data
    if plugin_registry is not None:
        _merge_plugin_overlay_profiles(
            profiles,
            plugin_registry=plugin_registry,
            plugin_configs=plugin_configs or {},
            warnings=warnings,
        )
    return profiles


def _merge_plugin_overlay_profiles(
    profiles: dict[str, dict[str, Any]],
    *,
    plugin_registry: PluginRegistry,
    plugin_configs: Mapping[str, Mapping[str, Any]],
    warnings: list[str] | None,
) -> None:
    for record in plugin_registry.compatible("overlay"):
        plugin_name = record.name
        if record.plugin is None:
            _warn_overlay(plugin_name, "has no plugin object and was skipped", warnings)
            continue
        context = PluginContext(
            sdk_version=plugin_registry.sdk_version,
            config=plugin_configs.get(plugin_name, {}),
        )
        try:
            raw_profile = record.plugin.load_overlay_profile(context)  # type: ignore[attr-defined]
        except Exception as exc:
            _warn_overlay(plugin_name, f"failed to load overlay profile: {exc}", warnings)
            continue
        profile = _validated_plugin_overlay_profile(plugin_name, raw_profile, warnings)
        if profile is None:
            continue
        overlay_id = profile["id"]
        if overlay_id in profiles:
            _warn_overlay(
                plugin_name,
                f"overlay id '{overlay_id}' collides with an existing overlay and was skipped",
                warnings,
            )
            continue
        profiles[overlay_id] = profile


def _validated_plugin_overlay_profile(
    plugin_name: str,
    raw_profile: Any,
    warnings: list[str] | None,
) -> dict[str, Any] | None:
    if not isinstance(raw_profile, Mapping):
        _warn_overlay(plugin_name, "overlay profile must be a mapping", warnings)
        return None

    profile = {str(key): value for key, value in raw_profile.items()}
    overlay_id = profile.get("id")
    if not isinstance(overlay_id, str) or not overlay_id.startswith("plugin:"):
        _warn_overlay(plugin_name, "overlay id must be explicit and namespaced as plugin:<name>", warnings)
        return None

    for field_name in ("name", "trigger_type"):
        if not isinstance(profile.get(field_name), str) or not profile[field_name].strip():
            _warn_overlay(plugin_name, f"overlay profile missing required '{field_name}' field", warnings)
            return None

    for field_name in ("triggered_by", "applicable_data_types"):
        value = profile.get(field_name)
        if value is not None and not _is_string_list(value):
            _warn_overlay(plugin_name, f"overlay profile field '{field_name}' must be a list of strings", warnings)
            return None

    for field_name in ("control_adjustments", "evidence_requirements", "controls"):
        value = profile.get(field_name)
        if value is not None and not isinstance(value, Mapping):
            _warn_overlay(plugin_name, f"overlay profile field '{field_name}' must be a mapping", warnings)
            return None

    return profile


def _is_string_list(value: Any) -> bool:
    return isinstance(value, list) and all(isinstance(item, str) for item in value)


def _warn_overlay(plugin_name: str, detail: str, warnings: list[str] | None) -> None:
    message = f"Skipping Ancilis plugin overlay {plugin_name}: {detail}"
    logger.warning(message)
    if warnings is not None:
        warnings.append(message)


def load_certification_profile(cert_id: str) -> dict[str, Any] | None:
    """Load a single certification profile by ID. Returns None if not found."""
    path = CERTIFICATIONS_DIR / f"{cert_id}.json"
    if not path.exists():
        logger.warning("Certification profile not found: %s", cert_id)
        return None
    data = json.loads(path.read_text())
    if "version" not in data:
        logger.warning("Certification profile '%s' missing required 'version' field", cert_id)
        return None
    return cast(dict[str, Any], data)


def load_certification_profiles(cert_ids: list[str]) -> dict[str, dict[str, Any]]:
    """Load multiple certification profiles by ID. Warns on missing profiles."""
    profiles: dict[str, dict[str, Any]] = {}
    for cid in cert_ids:
        profile = load_certification_profile(cid)
        if profile is not None:
            profiles[cid] = profile
    return profiles


def load_control_definitions() -> dict[str, dict[str, Any]]:
    """Load all control definition JSON files from shared/controls/."""
    controls: dict[str, dict[str, Any]] = {}
    for path in sorted(CONTROLS_DIR.glob("*.json")):
        data = json.loads(path.read_text())
        controls[data["id"]] = data
    return controls


def load_taxonomy() -> dict[str, Any]:
    """Load the classification taxonomy."""
    data = json.loads(CLASSIFICATIONS_FILE.read_text())
    return cast(dict[str, Any], data)
